In [0]:
from sklearn.naive_bayes import MultinomialNB, BernoulliNB, GaussianNB
from IPython.display import display, Markdown
import numpy as np
import pandas as pd

In [0]:
#--data as LISTS
sweet = [1,1,0,0,0]
data = [[3,4],[2,3],[0,3],[3,2],[1,4]]

#--data as Pandas dataframes
TraindfX = pd.DataFrame(dict(Color=[3,2,0,3,1],Weight=[4,3,3,2,4]))
TraindfY = pd.DataFrame(dict(Sweet=[1,1,0,0,0]))

#--inspect the List data
display(Markdown("### Inspect the List Data"))
print(sweet)
print(data)
print('Sweet <datatype>=' + str(type(sweet)))
print('Data <datatype>=' + str(type(data)))

### Inspect the List Data

[1, 1, 0, 0, 0]
[[3, 4], [2, 3], [0, 3], [3, 2], [1, 4]]
Sweet <datatype>=<class 'list'>
Data <datatype>=<class 'list'>


In [0]:
display(Markdown("### Inspect the Pandas DataFrame Data"))
print('TraindfY <datatype>=' + str(type(TraindfY)))
print('TraindfX <datatype>=' + str(type(TraindfX)))
display(TraindfY)
display(TraindfX)

### Inspect the Pandas DataFrame Data

TraindfY <datatype>=<class 'pandas.core.frame.DataFrame'>
TraindfX <datatype>=<class 'pandas.core.frame.DataFrame'>


Sweet
0      1
1      1
2      0
3      0
4      0

Color  Weight
0      3       4
1      2       3
2      0       3
3      3       2
4      1       4

In [0]:
#--Create a Python dictionary with cross-tabulation reports for each predictor variable and the outcome variable showing COUNTS and PERCENTS
SweetDataCrosstab = {'SweetColorCrosstab':{'Count':pd.crosstab(TraindfY.Sweet,TraindfX.Color,margins=True),'Percent':pd.crosstab(TraindfY.Sweet,TraindfX.Color,normalize='index')},
                     'SweetWeightCrosstab': {'Count':pd.crosstab(TraindfY.Sweet,TraindfX.Weight,margins=True),'Percent':pd.crosstab(TraindfY.Sweet,TraindfX.Weight,normalize='index')}                    
                   }
display(Markdown('## Display Apriori values and Cross-tab Statistics for Data'))
display(Markdown('### Apriori for Sweet/!Sweet'))
display(Markdown('Sweet% = ' + str(SweetDataCrosstab['SweetColorCrosstab']['Count'].values[1][4]/SweetDataCrosstab['SweetColorCrosstab']['Count'].values[2][4])))
display(Markdown('!Sweet% = ' + str(SweetDataCrosstab['SweetColorCrosstab']['Count'].values[0][4]/SweetDataCrosstab['SweetColorCrosstab']['Count'].values[2][4])))

display(Markdown("# Count and PCT for Sweet by Color"))
display(Markdown("### Count"),SweetDataCrosstab['SweetColorCrosstab']['Count'],Markdown("### Percent"),SweetDataCrosstab['SweetColorCrosstab']['Percent'])
display(Markdown("# Count and PCT for Sweet by Weight"))
display(Markdown("### Count"),SweetDataCrosstab['SweetWeightCrosstab']['Count'],Markdown("### Percent"),SweetDataCrosstab['SweetWeightCrosstab']['Percent'])

## Display Apriori values and Cross-tab Statistics for Data

### Apriori for Sweet/!Sweet

Sweet% = 0.4

!Sweet% = 0.6

# Count and PCT for Sweet by Color

### Count

Color  0  1  2  3  All
Sweet                 
0      1  1  0  1    3
1      0  0  1  1    2
All    1  1  1  2    5

### Percent

Color         0         1    2         3
Sweet                                   
0      0.333333  0.333333  0.0  0.333333
1      0.000000  0.000000  0.5  0.500000

# Count and PCT for Sweet by Weight

### Count

Weight  2  3  4  All
Sweet               
0       1  1  1    3
1       0  1  1    2
All     1  2  2    5

### Percent

Weight         2         3         4
Sweet                               
0       0.333333  0.333333  0.333333
1       0.000000  0.500000  0.500000

In [0]:
#SweetDataCrosstab

In [0]:
#--initialize the "classifier" from SciKit-Learn to use one of three models
#classifier = GaussianNB()
#classifier = MultinomialNB()
classifier = BernoulliNB()

#--execute the classifier to create a model based on the data provided
classifier = classifier.fit(data,sweet)

display("===== BASIC MODEL OUTPUT =====")
display("Score")
display(classifier.score(data,sweet))
display("params")
display(classifier.get_params(deep=True))
display("calulated predicted probabilities")
display(classifier.predict_proba(data))

'===== BASIC MODEL OUTPUT ====='

'Score'

0.6

'params'

{'alpha': 1.0, 'binarize': 0.0, 'class_prior': None, 'fit_prior': True}

'calulated predicted probabilities'

array([[0.56140351, 0.43859649],
       [0.56140351, 0.43859649],
       [0.71910112, 0.28089888],
       [0.56140351, 0.43859649],
       [0.56140351, 0.43859649]])

In [0]:
#--Create TEST data and test the Model
ptestdata = [[3,4],[1,4],[3,4],[1,4],[1,4]]

#--Execute the classifer prediction using the TEST data
predicted = classifier.predict(ptestdata)
print(predicted)

#--import the metrics library and generate report
from sklearn import metrics
print(metrics.classification_report(sweet,predicted)) 
print(metrics.confusion_matrix(sweet, predicted))
print(metrics.accuracy_score(sweet, predicted))

[0 0 0 0 0]
             precision    recall  f1-score   support

          0       0.60      1.00      0.75         3
          1       0.00      0.00      0.00         2

avg / total       0.36      0.60      0.45         5

[[3 0]
 [2 0]]
0.6


C:\Program Files (x86)\Microsoft Visual Studio\Shared\Anaconda3_64\lib\site-packages\sklearn\metrics\classification.py:1135: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


In [0]:
#--Using Pandas Dataframs for analysis

classifier2 = GaussianNB()
#classifier2 = MultinomialNB()
#classifier2 = BernoulliNB()

classifier2.fit(TraindfX,TraindfY)

print("Score")
print(classifier2.score(TraindfX,TraindfY))
print("params")
print(classifier2.get_params(deep=True))
print("Predict Proba")
print(classifier2.predict_proba(TraindfX).round(decimals=2))
print("Predict LOG Proba")
print(classifier2.predict_log_proba(TraindfX).round(decimals=2))

Score
1.0
params
{'priors': None}
Predict Proba
[[0.16 0.84]
 [0.46 0.54]
 [1.   0.  ]
 [0.91 0.09]
 [0.96 0.04]]
Predict LOG Proba
[[ -1.82  -0.18]
 [ -0.77  -0.62]
 [ -0.   -11.43]
 [ -0.09  -2.45]
 [ -0.04  -3.25]]


C:\Program Files (x86)\Microsoft Visual Studio\Shared\Anaconda3_64\lib\site-packages\sklearn\utils\validation.py:578: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [0]:
#Predict Output
ptestdata2 = [[3,4],[1,4],[3,4],[1,4],[1,4]]
ptestdatadf2 = pd.DataFrame(dict(Color=[3,2,0,3,1],Weight=[4,3,3,2,4]))
#print(ptestdatadf2)

predicted2 = classifier2.predict(ptestdata2)
print(predicted2)

from sklearn import metrics
print(metrics.classification_report(TraindfY,predicted2)) 

print(metrics.accuracy_score(TraindfY, predicted2))

CM = metrics.confusion_matrix(TraindfY, predicted2)
print(CM)
#import matplotlib.pyplot as plt
#plt.imshow(CM, cmap= 'binary')

[1 0 1 0 0]
             precision    recall  f1-score   support

          0       0.67      0.67      0.67         3
          1       0.50      0.50      0.50         2

avg / total       0.60      0.60      0.60         5

0.6
[[2 1]
 [1 1]]


--Naive Bayes Model: Application

Naive Bayes Model use conditional probabilities to predict the label value. Models are naive since they assume statistical independence of features. Adavantages of this model are that it work well with minimal training data, ensure regularization, and it is computationally efficient and highly scalable. 
Naive Bayes models are used in many situations including:
Document classification
SPAM detection
Image classification,etc

---

From the above case, A-priori probabilities tell us that we can tell 60% of the time from the train data that the apple is sweet and 40% of the time that it is not sweet. We then look at the conditional probabilities. Given the color of the apple (color type 0 to 3), the conditional probability that the apple is sweet is displayed.We can say there is 50-50% chace that the apple is not sweet when it has type 2 or type 3 color. Similarly, given the weight of type (2-4), the conditional probability of sweet/non-sweet is calculated.

We use BernoulliNB() function on the trainining dataset since we are dealing with binary distribution funcrion. classifier2.score() internally calculates Y'=predictor.predict(X) and then compares Y' against Y to give an accuracy measure. The score of 60% is not ideal but acceptable. 
We then predict the probability of each instance values for both classes. The first index refers to the probability that the data belong to class 0, and the second refers to the probability that the data belong to class 1. 
After training the data, we use test data to analyze and identify whether our model is working properly. Precision in Classification report in calculated as (TruePositive/(TruePositive+FalsePositive)) whereas Recall = (TP/(TP+FN)). In words, Precisions is the numbers of correct positive results given the number of positive results predicted by classifier. On the other hand, Recall is the number of correct positive results given the number of all relevent sample that should have been identified as positive). There is always a trade off between these two indicators and F1 score reconciled them. (Support is the number of true sample that lies in that class) In our case we have better precision score than recall value ( 0.67 vs 0.5) but overall f1 score of 0.6 is still an okay indicator.
The confusion Matrix then tell us the number of samples in each categories of True Negative, True Positive, False Positive, False Negative. The diagonal elements of Confustion matrix represent the number of points for which the predicted label is equal to the true label. Out of 5 samples, 2 actaul'0' matches with the predicted '0' and only 1 actualy '1' matches with the predicted '1'. 